### PR TITLE
core: Expose InsertPoint class (NFC)

### PR DESCRIPTION
**Pull request by math-fehr**
This PR creates the InsertPoint class, and use it in the builder.
This has no functional changes otherwise

### CODE REVIEW

File: xdsl/builder.py [hunk 1]
```diff
@@ -12,47 +12,92 @@
...
+    def __init__(self, point: Block) -> None:
+        ...
+
+    @overload
+    def __init__(self, point: Block, point2: Operation | None) -> None:
+        ...
+
+    def __init__(
+        self, point: Block | Operation, point2: Operation | None = None
+    ) -> None:
```
Comment (webmiche) [inline xdsl/builder.py:46]: Could this be `{inner,outer}_point` or something? `point{1,2}` seems like suboptimal naming :wink:

**Review by webmiche — APPROVED** [review]

File: xdsl/builder.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/builder.py:46]: I have a feeling that it can't be due to Python's naming things, but I would suggest `op_before_or_block, op_before`

**Review by PapyChacal — APPROVED** [review]
Thanks! I love this but I'm not sure why, given it's a NFC. OOC, can I ask about your concrete motivation for implementing it, if any?

**Comment by math-fehr** [comment]
> Thanks! I love this but I'm not sure why, given it's a NFC. OOC, can I ask about your concrete motivation for implementing it, if any?

I'm currently thinking of changing the rewriter, for mostly two reasons:
* One is that Dalia would need one that works with a worklist, and currently we only support rewrites that operate in order in the IR.
* The other is that I'm tired of writing `rewriter.insert_op_before_matched_op`. I just want to use the builder, so I first need to make `RewritePattern` inherit from `Builder`. But while looking at it I felt that `InsertPoint` should be its own class. I might never use it anywhere else, but I felt it's better that way.

**Pull request merged**